File: solutions/200.number-of-islands/disjoint-set.ts
/*
 * @lc app=leetcode.cn id=200 lang=typescript
 *
 * https://leetcode-cn.com/problems/number-of-islands
 */

// @lc code=start
class DisjointSet {
    private allies: number[] = [];
    private count: number = 0;

    constructor(grid: string[][]) {
        const rowNum = grid.length;
        const colNum = grid[0].length;
        for (let row = 0; row < rowNum; ++row) {
            for (let col = 0; col < colNum; ++col) {
                if (grid[row][col] === '1') {
                    this.allies.push(row * colNum + col);
                    this.count += 1;
                } else {
                    this.allies.push(-1);
                }
            }
        }
    }

    find(i: number): number {
        if (this.allies[i] !== i) {
            // compress path
            this.allies[i] = this.find(this.allies[i]);
        }
        return this.allies[i];
    }

    unite(x: number, y: number): void {
        const rootx = this.find(x);
        const rooty = this.find(y);
        if (rootx === rooty) {
            return;
        }
        this.allies[rooty] = rootx;
        this.count -= 1;
    }

    getCount(): number {
        return this.count;
    }
}

function numIslands(grid: string[][]): number {
    if (grid.length === 0 || grid[0].length === 0) {
        return 0;
    }
    const rowNum = grid.length;
    const colNum = grid[0].length;
    const disjointSet = new DisjointSet(grid);
    for (let row = 0; row < rowNum; ++row) {
        for (let col = 0; col < colNum; ++col) {
            if (grid[row][col] === '0') {
                continue;
            }
            // down
            if (row + 1 < rowNum && grid[row + 1][col] === '1') {
                disjointSet.unite(row * colNum + col, (row + 1) * colNum + col);
            }
            // right
            if (col + 1 < colNum && grid[row][col + 1] === '1') {
                disjointSet.unite(row * colNum + col, row * colNum + col + 1);
            }
        }
    }
    return disjointSet.getCount();
}
// @lc code=end

export {};

File: solutions/200.number-of-islands/src/disjoint_set.rs
/*
 * @lc app=leetcode.cn id=200 lang=rust
 *
 * https://leetcode.cn/problems/number-of-islands/
 */

pub struct Solution;

// @lc code=start
struct DisjointSet {
    parent: Vec<usize>,

    count: i32,
}

impl DisjointSet {
    pub fn new(grid: &Vec<Vec<char>>) -> Self {
        let rows = grid.len();
        let cols = grid[0].len();
        let mut parent = Vec::with_capacity(rows * cols);
        let mut count = 0;
        for i in 0..rows {
            for j in 0..cols {
                parent.push(i * cols + j);
                if grid[i][j] == '1' {
                    count += 1;
                }
            }
        }
        DisjointSet { parent, count }
    }

    pub fn find(&mut self, x: usize) -> usize {
        if self.parent[x] != x {
            self.parent[x] = self.find(self.parent[x]);
        }
        self.parent[x]
    }

    pub fn union(&mut self, x: usize, y: usize) {
        let parent_x = self.find(x);
        let parent_y = self.find(y);
        if parent_x == parent_y {
            return;
        }
        self.parent[parent_y] = parent_x;
        self.count -= 1;
    }

    pub fn get_count(&self) -> i32 {
        self.count
    }
}

impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let rows = grid.len();
        let cols = grid[0].len();
        let mut set = DisjointSet::new(&grid);
        for i in 0..rows {
            for j in 0..cols {
                if grid[i][j] == '0' {
                    continue;
                }
                if i + 1 < rows && grid[i + 1][j] == '1' {
                    set.union(i * cols + j, (i + 1) * cols + j);
                }
                if j + 1 < cols && grid[i][j + 1] == '1' {
                    set.union(i * cols + j, i * cols + j + 1);
                }
            }
        }
        set.get_count()
    }
}
// @lc code=end
